compiler/src/dotty/tools/dotc/core/TyperState.scala

```diff
@@ -89,6 +90,14 @@ class TyperState() {
   def ownedVars: TypeVars = myOwnedVars
   def ownedVars_=(vs: TypeVars): Unit = myOwnedVars = vs
 
+  /** The set of type variables `tv` such that, if `tv` is instantiated to
+   *  its lower bound, top-level soft unions in the instance type are converted
+   *  to hard unions instead of being widened in `widenOr`.
+   */
+  private var myRigidVars: TypeVars = _
+  def rigidVars: TypeVars = myRigidVars
```

Comment:
I like this much more :sweat_smile:, but I think the name could be misleading: usually "rigid variables" are variables that cannot be constrained at all (like the type parameters of a method as seen from inside that method).  Maybe go with `hardVars` instead in keeping with `hardenVars`?